File: hmh-frontend/src/api/lotTypes.ts
import client from "./client";

export interface LotType {
  id:                     string;
  project_id:             string;
  name:                   string;
  code:                   string | null;
  description:            string | null;
  default_template_id:    string | null;
  default_template_name:  string | null;
  lot_count:              number;
  created_at:             string;
  updated_at:             string;
}

export interface LotTypeLotSummary {
  id:            string;
  lot_number:    string;
  unit_type:     string | null;
  site_id:       string | null;
  status:        string;
  boq_customized: boolean;
}

export interface LotTypeWithLots extends LotType {
  lots: LotTypeLotSummary[];
}

export interface LotTypeCreate {
  name:                string;
  code?:               string | null;
  description?:        string | null;
  default_template_id?: string | null;
}

export interface LotTypeUpdate {
  name?:               string;
  code?:               string | null;
  description?:        string | null;
  default_template_id?: string | null;
}

export interface AssignResult {
  lot_type_id:   string;
  lot_type_name: string;
  assigned:      number;
  reassigned:    number;
  already_set:   number;
  total:         number;
}

export interface RemoveResult {
  lot_type_id:   string;
  lot_type_name: string;
  removed:       number;
  not_in_type:   number;
}

export const lotTypesApi = {
  list: async (projectId: string): Promise<LotType[]> => {
    const res = await client.get<{ data: LotType[] }>(
      `/projects/${projectId}/lot-types/`
    );
    return res.data.data;
  },

  create: async (projectId: string, body: LotTypeCreate): Promise<LotType> => {
    const res = await client.post<{ data: LotType }>(
      `/projects/${projectId}/lot-types/`,
      body
    );
    return res.data.data;
  },

  get: async (lotTypeId: string): Promise<LotTypeWithLots> => {
    const res = await client.get<{ data: LotTypeWithLots }>(
      `/lot-types/${lotTypeId}`
    );
    return res.data.data;
  },

  update: async (lotTypeId: string, body: LotTypeUpdate): Promise<LotType> => {
    const res = await client.patch<{ data: LotType }>(
      `/lot-types/${lotTypeId}`,
      body
    );
    return res.data.data;
  },

  delete: async (lotTypeId: string): Promise<void> => {
    await client.delete(`/lot-types/${lotTypeId}`);
  },

  assignLots: async (lotTypeId: string, lotIds: string[]): Promise<AssignResult> => {
    const res = await client.post<{ data: AssignResult }>(
      `/lot-types/${lotTypeId}/assign-lots`,
      { lot_ids: lotIds }
    );
    return res.data.data;
  },

  removeLots: async (lotTypeId: string, lotIds: string[]): Promise<RemoveResult> => {
    const res = await client.post<{ data: RemoveResult }>(
      `/lot-types/${lotTypeId}/remove-lots`,
      { lot_ids: lotIds }
    );
    return res.data.data;
  },
};

File: hmh-backend/app/api/v1/lot_types.py
"""LotType routes — Phase 3D.2."""

import uuid

from fastapi import APIRouter

from app.dependencies import ALL_ROLES, CurrentUser, DbSession, OFFICE_AND_ABOVE
from app.schemas.common import ApiSuccess
from app.schemas.lot_type import (
    AssignLotsRequest,
    LotTypeCreate,
    LotTypeRead,
    LotTypeUpdate,
    LotTypeWithLots,
    RemoveLotsRequest,
)
from app.services import lot_type_service

project_lot_types_router = APIRouter(
    prefix="/projects/{project_id}/lot-types",
    tags=["lot-types"],
)
lot_types_router = APIRouter(prefix="/lot-types", tags=["lot-types"])


@project_lot_types_router.get(
    "/",
    response_model=ApiSuccess[list[LotTypeRead]],
    dependencies=[ALL_ROLES],
)
def list_lot_types(project_id: uuid.UUID, db: DbSession):
    """List all lot types for a project with lot counts."""
    types = lot_type_service.list_lot_types(db, project_id)
    return ApiSuccess(data=types)


@project_lot_types_router.post(
    "/",
    response_model=ApiSuccess[LotTypeRead],
    status_code=201,
    dependencies=[OFFICE_AND_ABOVE],
)
def create_lot_type(
    project_id: uuid.UUID,
    body: LotTypeCreate,
    db: DbSession,
    current_user: CurrentUser,
):
    lt = lot_type_service.create_lot_type(db, project_id, body, actor_id=current_user.id)
    return ApiSuccess(data=lt, message=f"Lot type '{lt.name}' created.")


@lot_types_router.get(
    "/{lot_type_id}",
    response_model=ApiSuccess[LotTypeWithLots],
    dependencies=[ALL_ROLES],
)
def get_lot_type(lot_type_id: uuid.UUID, db: DbSession):
    """Get a lot type with its linked lots."""
    lt = lot_type_service.get_lot_type(db, lot_type_id)
    return ApiSuccess(data=lt)


@lot_types_router.patch(
    "/{lot_type_id}",
    response_model=ApiSuccess[LotTypeRead],
    dependencies=[OFFICE_AND_ABOVE],
)
def update_lot_type(lot_type_id: uuid.UUID, body: LotTypeUpdate, db: DbSession):
    lt = lot_type_service.update_lot_type(db, lot_type_id, body)
    return ApiSuccess(data=lt, message=f"Lot type '{lt.name}' updated.")


@lot_types_router.delete(
    "/{lot_type_id}",
    response_model=ApiSuccess[dict],
    dependencies=[OFFICE_AND_ABOVE],
)
def delete_lot_type(lot_type_id: uuid.UUID, db: DbSession):
    """
    Delete a lot type. Blocked if any lots are still assigned.
    Unassign all lots first, or use the force query param.
    """
    from app.models.lot_type import LotType
    lt = db.get(LotType, lot_type_id)
    name = lt.name if lt else str(lot_type_id)
    lot_type_service.delete_lot_type(db, lot_type_id)
    return ApiSuccess(data={"id": str(lot_type_id)}, message=f"Lot type '{name}' deleted.")


# ── Lot assignment ────────────────────────────────────────────────────────────

@lot_types_router.post(
    "/{lot_type_id}/assign-lots",
    response_model=ApiSuccess[dict],
    dependencies=[OFFICE_AND_ABOVE],
)
def assign_lots(lot_type_id: uuid.UUID, body: AssignLotsRequest, db: DbSession):
    """
    Assign lots to this type (bulk).
    Lots from other types are reassigned here.
    Returns: { assigned, reassigned, already_set, total }
    """
    result = lot_type_service.assign_lots(db, lot_type_id, body.lot_ids)
    return ApiSuccess(
        data=result,
        message=(
            f"{result['assigned']} lot(s) assigned, "
            f"{result['reassigned']} reassigned, "
            f"{result['already_set']} already in type."
        ),
    )


@lot_types_router.post(
    "/{lot_type_id}/remove-lots",
    response_model=ApiSuccess[dict],
    dependencies=[OFFICE_AND_ABOVE],
)
def remove_lots(lot_type_id: uuid.UUID, body: RemoveLotsRequest, db: DbSession):
    """Remove lots from this type (sets lot_type_id = NULL)."""
    result = lot_type_service.remove_lots(db, lot_type_id, body.lot_ids)
    return ApiSuccess(data=result, message=f"{result['removed']} lot(s) removed from type.")
